drivers/i2c/i2c_npcx_controller.c

```diff
@@ -0,0 +1,878 @@
+ *                              |       No        |                       | Yes
+ *    +------+   +------------+ |  +------- ----+ |    +------- -------+  |
+ * +->| IDLE |-->| WAIT_START |--->| WRITE_FIFO |-+--->| WRITE_SUSPEND |--+
+ * |  +------+   +------------+    +------------+  Yes +---------------+  |
+ * |     Issue START     START completed                                  | No
+ * |      +---------------------------------------------------------------+
+ * |      |
+ * |      |                 All msg data received?  Is there next read msg?
+ * |      |                   +<-----------------+<-----------------------+
+ * |      |                   |       No         |                        | Yes
+ * |      |  +--------------+ |  +------- ---+   |    +------- ------+    |
+ * |      +--| WAIT_RESTART |--->| READ_FIFO |---+--->| READ_SUSPEND |----+
+ * |         +--------------+    +-----------+    Yes +--------------+    |
+ * |  Issue RESTART      RESTART completed                                | No
+ * |                                           +-----------+              |
+ * +-------------------------------------------| WAIT_STOP |<-------------+
+ *              STOP is completed              +-----------+  Issue STOP
+ *
+ */
+
+#include <assert.h>
+#include <drivers/i2c.h>
+#include <drivers/clock_control.h>
+#include <soc.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(i2c_npcx, LOG_LEVEL_ERR);
+
+/* I2C bus frequency */
+#define NPCX_I2C_BUS_SPEED_100KHZ 0
+#define NPCX_I2C_BUS_SPEED_400KHZ 1
+#define NPCX_I2C_BUS_SPEED_1MHZ   2
+
+/* I2C controller mode */
+#define NPCX_I2C_BANK_NORMAL 0
+#define NPCX_I2C_BANK_FIFO   1
+
+/* Timeout for device should be available after reset (SMBus spec. unit:ms) */
+#define I2C_MAX_TIMEOUT 35
+
+/* Timeout for SCL held to low by slave device . (SMBus spec. unit:ms). */
+#define I2C_MIN_TIMEOUT 25
+
+/* Default maximum time we allow for an I2C transfer (unit:ms) */
+#define I2C_TRANS_TIMEOUT K_MSEC(100)
+
+/*
+ * NPCX I2C module that supports FIFO mode has 32 bytes Tx FIFO and
+ * 32 bytes Rx FIFO.
+ */
+#define NPCX_I2C_FIFO_MAX_SIZE 32
+
+/* Valid bit fields in SMBST register */
+#define NPCX_VALID_SMBST_MASK ~(BIT(NPCX_SMBST_XMIT) | BIT(NPCX_SMBST_MASTER))
+
+/*
+ * Internal SMBus Interface driver states values, which reflect events
+ * which occurred on the bus
+ */
+enum smb_oper_state {
+	NPCX_I2C_IDLE,
+	NPCX_I2C_WAIT_START,
+	NPCX_I2C_WAIT_RESTART,
+	NPCX_I2C_WRITE_FIFO,
+	NPCX_I2C_WRITE_SUSPEND,
+	NPCX_I2C_READ_FIFO,
+	NPCX_I2C_READ_SUSPEND,
+	NPCX_I2C_WAIT_STOP,
+	NPCX_I2C_ERROR_RECOVERY,
+};
+
+/* I2C timing configuration for each i2c speed */
+struct npcx_i2c_timing_cfg {
+	uint8_t HLDT; /* i2c hold-time (Unit: clocks) */
+	uint8_t k1; /* k1 = SCL low-time (Unit: clocks) */
+	uint8_t k2; /* k2 = SCL high-time (Unit: clocks) */
+};
+
+/* Device config */
+struct i2c_ctrl_config {
+	uintptr_t base; /* i2c controller base address */
```

Comment:
I recall seeing this before; is there a reason you don't cast this to the pointer type when it's stored here, rather than doing it in the `HAL*INSTANCE` macros?  If you've standardized on storing `uintptr_t` better to keep it consistent, otherwise this looks weird.

Reply:
We have a discussion [here](https://github.com/zephyrproject-rtos/zephyr/pull/27578#discussion_r476550183). The main reason (it's also our burden) is the layout of partial module registers is irregular. The clock control registers' layout in PMC is a good example. The register offset of PWDWN_CTL1-6 is [0x08-0x0D](https://github.com/zephyrproject-rtos/zephyr/blob/master/soc/arm/nuvoton_npcx/common/reg/reg_def.h#L116)  But the register offset of PWDWN_CTL7 is [0x24](https://github.com/zephyrproject-rtos/zephyr/blob/master/soc/arm/nuvoton_npcx/common/reg/reg_def.h#L122) For this case, our HAL uses macro functions to handle [this](https://github.com/zephyrproject-rtos/zephyr/blob/master/soc/arm/nuvoton_npcx/common/reg/reg_def.h#L126). Here is an usage [example](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/clock_control/clock_control_npcx.c#L55) 

That's why I keep "uintptr_t " type and cast it in HAL macro if we need accessing register via structure. My plan is keeping this style temporarily. Once the other drivers are landed (There're too many similar macro functions in our HAL.), I will submit a refactor RR, and move these macros to inline functions. Then we can replace them with the pointer type of their structure.

Reply:
OK with me.